structured_prediction_baselines/modules/sampler/sequence_tagging/sequence_tagging_inference.py

```diff
@@ -0,0 +1,224 @@
+            score_nn=score_nn, oracle_value_function=oracle_value_function
+        )
+        trainable_parameters = {}
+        trainable_parameters.update(inference_nn.named_parameters())
+        trainable_parameters.update(cost_augmented_layer.named_parameters())
+        optimizer_ = optimizer.construct(model_parameters=list(trainable_parameters.items()))
+
+        return cls(
+            optimizer_,
+            loss_fn_,
+            inference_nn=inference_nn,
+            score_nn=score_nn,
+            cost_augmented_layer=cost_augmented_layer,
+            oracle_value_function=oracle_value_function,
+            stopping_criteria=stopping_criteria,
+        )
+
+    def get_loss_fn(
+        self, x: Any, labels: Optional[torch.Tensor]
+    ) -> Callable[[torch.Tensor], torch.Tensor]:
+        # Sampler gets labels of shape (batch, ...), hence this
+        # function will get labels of shape (batch*num_init_samples, ...)
+        # but Loss expect y or shape (batch, num_samples or 1, ...)
+        # Also during eval the loss is different. We inform the loss function
+        # using None
+
+        if self.training and (labels is None):
+            warnings.warn("Labels should not be None in training mode!")
+
+        def loss_fn(y_hat: torch.Tensor, y_cost_aug: torch.Tensor, buffer: Dict) -> torch.Tensor:
+            return self.loss_fn(
+                x,
+                (
+                    labels  # E:labels.unsqueeze(1)
+                    if (self.training and labels is not None)
+                    else None
+                ),
+                y_hat,
+                y_cost_aug,
+                buffer,
+                None,
+            )
+
+        return loss_fn
+
+    def get_dtype_device(self) -> Tuple[torch.dtype, torch.device]:
+        for param in self.loss_fn.parameters():
+            dtype = param.dtype
+            device = param.device
+
+            break
+
+        return dtype, device
+
+    def get_batch_size(self, x: Any) -> int:
+        if isinstance(x, torch.Tensor):
+            return x.shape[0]
+        else:
+            raise NotImplementedError
+
+    @contextlib.contextmanager
+    def param_grad(self) -> Generator[None, None, None]:
+        trainable_modules = [self.inference_nn, self.cost_augmented_layer]
+        requires_grad_map = {}
+        for module in trainable_modules:
+            if module is not None:
+                try:
+                    for param in module.parameters():
+                        param.requires_grad = True
+                    yield
+                finally:
+                    # set the requires_grad back to false
+                    for param in module.parameters():
+                        param.requires_grad = False
+            else:  # if there is no loss_fn, we have nothing to do.
+                warnings.warn(
+                    (
+                        "There is inference_nn in gradient based inference sampler."
```

Comment:
Change this warning to a more helpful one like: "You are using inference network-based model but cost_augmented_layer and interence_nn are both None. Either you are not using the right Sampler or you have not constructed it correctly."